### ISSUE TITLE
typescript; no-used-vars fires incorrectly on identifiers that are in use

### ISSUE THREAD

**Issue by plastikfan** [eslint/eslint#12963]
**Tell us about your environment**

Node version: v12.13.0
npm version: v6.14.0
Local ESLint version: v6.8.0 (Currently used)
Global ESLint version: Not found

**What parser (default, Babel-ESLint, etc.) are you using?**
'@typescript-eslint/parser'
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    node: true,
    browser: true,
    es6: true,
    mocha: true
  },
  extends: [
    'semistandard'
  ],
  globals: {
    Atomics: 'readonly',
    SharedArrayBuffer: 'readonly'
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 2018, 
    sourceType: 'module'
  },
  plugins: [
    '@typescript-eslint'
  ],
  rules: {
  }
}

```
Please also note that I'm using semistandard as can be seen in the "extends" property above.

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js

import * as R from 'ramda';
import * as jaxom from 'jaxom-ts';
import { functify } from 'jinxed';
// eslint-disable-next-line no-unused-vars
import * as types from '../types';

export class ExpressionBuilderImpl {
  // eslint-disable-next-line no-useless-constructor
  constructor (private converter: jaxom.IConverter, private specSvc: jaxom.ISpecService,
    private parseInfo: jaxom.IParseInfo, private xpath: types.ISelectors) { }

  public buildExpressionGroup (parentNode: Node, groupName: string)
    : types.StringIndexableObj {
    const expressionsGroupNode = this.xpath.selectById(
      'Expressions', 'name', groupName, parentNode);

```
As you can see the rule has been disabled for the so called offending line, but I shouldn't have to do this for this scenario.

I have found this rule firing incorrectly in many places. Another example:

```js
import { IExecutionContext, ICommandExecutionResult } from './cli-types';
import * as factory from './commands/command-factory';
export function run (executionContext: IExecutionContext)
: number {
  const command = factory.construct(executionContext);
  const executionResult: ICommandExecutionResult = command.exec(executionContext);
  return executionResult.resultCode;
}
```
 Both IExecutionContext and ICommandExecutionResult are both reported as being unused, but they are indeed both referenced in the "run" method signature and within the implementation.

```bash
npm run lint
```

where "lint" is defined as:

```bash
npx eslint -c .eslintrc.js "./lib/**/*.{js,ts,tsx}" "./tests/**/*.{js,ts,tsx}" --fix
```
**What did you expect to happen?**
Rule to be applied as documented and work for typescript too. However, I recognise that the development stack is large and complex and that this may be an issue with something else; ie the parser: '@typescript-eslint/parser', so your advice will be appreciated. Since tslint has been deprecated in favour os eslint, I expect eslint to apply the rules just as sucessfully as tslint does.

**What actually happened? Please include the actual, raw output from ESLint.**

zenobia-ts@0.0.1 fix: `npx eslint -c .eslintrc.js "./lib/**/*.{js,ts,tsx}" "./tests/**/*.{js,ts,tsx}" --fix`

```
/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/application.ts
  1:10  error  'IExecutionContext' is defined but never used        no-unused-vars
  1:29  error  'ICommandExecutionResult' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/cli-types.ts
  1:13  error  'jaxom' is defined but never used                   no-unused-vars
  2:10  error  'VirtualFS' is defined but never used               no-unused-vars
  2:21  error  'ICommandBuilderFactory' is defined but never used  no-unused-vars
  2:45  error  'ISelectors' is defined but never used              no-unused-vars
  2:57  error  'StringIndexableObj' is defined but never used      no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/command-line.class.ts
  2:13  error  'yargs' is defined but never used  no-unused-vars
  4:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/commands/cli-command.class.ts
  2:13  error  'jaxom' is defined but never used  no-unused-vars
  3:13  error  'ct' is defined but never used     no-unused-vars
  5:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/commands/command-factory.ts
   1:13  error  'ct' is defined but never used  no-unused-vars


/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/commands/jax-command.class.ts
  3:13  error  'jaxom' is defined but never used  no-unused-vars
  6:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/cli/composition-root.ts
   4:13  error  'types' is defined but never used                            no-unused-vars
   5:10  error  'IExecutionContext' is defined but never used                no-unused-vars
   5:29  error  'ICommandLineInputs' is defined but never used               no-unused-vars
   6:10  error  'IApplicationConsole' is defined but never used              no-unused-vars
  11:8   error  'yargs' is defined but never used                            no-unused-vars


/Users/Plastikfan/dev/github/js/zenobia-ts/lib/declarations.d.ts
  8:12  error  'patchFs' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/regex/expression-builder.class.ts
  4:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/regex/expression-builder.impl.ts
   5:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/types.ts
  2:13  error  'memfs' is defined but never used  no-unused-vars
  3:13  error  'jaxom' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/utils/helpers.ts
  3:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/zen-cli/builders/command-builder-factory.ts
  1:13  error  'jaxom' is defined but never used  no-unused-vars
  2:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/zen-cli/builders/command-builder.class.ts
  2:13  error  'jaxom' is defined but never used  no-unused-vars
  3:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/lib/zen-cli/builders/command-builder.impl.ts
  3:13  error  'jaxom' is defined but never used  no-unused-vars
  4:13  error  'types' is defined but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/tests/cli/application.class.spec.ts
   2:10  error  'functify' is defined but never used                         no-unused-vars
   6:13  error  'memfs' is defined but never used                            no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/tests/cli/command-line.class.spec.ts
   2:10  error  'functify' is defined but never used              no-unused-vars
   4:13  error  'memfs' is defined but never used                 no-unused-vars
  38:13  error  'inputs' is assigned a value but never used       no-unused-vars
  52:13  error  'inputs' is assigned a value but never used       no-unused-vars
  64:9   error  'zenobiaArgv' is assigned a value but never used  no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/tests/cli/commands/jax-command.class.spec.ts
   5:13  error  'R' is defined but never used                            no-unused-vars
   6:13  error  'fs' is defined but never used                           no-unused-vars
   7:13  error  'types' is defined but never used                        no-unused-vars
  31:7   error  'applicationConsole' is assigned a value but never used  no-unused-vars


/Users/Plastikfan/dev/github/js/zenobia-ts/tests/regex/expression-builder.class.spec.ts
   2:10  error  'functify' is defined but never used                         no-unused-vars
   8:13  error  'types' is defined but never used                            no-unused-vars


/Users/Plastikfan/dev/github/js/zenobia-ts/tests/regex/expression-builder.impl.spec.ts
    9:13  error  'types' is defined but never used                            no-unused-vars

/Users/Plastikfan/dev/github/js/zenobia-ts/tests/regex/expression-builder.test-config.spec.ts
   1:10  error  'functify' is defined but never used                         no-unused-vars
  
/Users/Plastikfan/dev/github/js/zenobia-ts/tests/zen-cli/builders/command-builder.spec.ts
   9:13  error  'types' is defined but never used                            no-unused-vars
```

Another situation in which the rule fires incorrectly is with ambient declarations:
I'm using fs-monkey to patch my use of the fs standard library in particular, I'm using it's patchFs function.
decalarations.d.ts:

```
declare module 'fs-monkey' {
  function patchFs(vol: any, fs: any);
}
```
"patchFs" is reported as being unused, but if this was removed, then my project wouldn't build successfully.

This rule seems to be firing in an unconsistent manner so I'v no idea what could be causing this issue. An import in one module can be exactly the same in a different module, but the firing of the rule can be different.

**Are you willing to submit a pull request to fix this bug?**
I don't have the expertise to submit a pull request. I am however will to assist in doing more investigation and helping in any way I can.

**Comment by mdjermanovic** [comment]
Hi @plastikfan, thanks for the issue!

I believe that [@typescript-eslint/no-unused-vars](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md) rule from [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin) works better than the core no-unused-vars on typescript source files.

**Comment by plastikfan** [comment]
Thanks.
